File: CITS3002 - Networks/Project - Transport Station - Client and Server/C++Project/BrowserHandler.cpp

#include "globals.h"

/**
 * Initialize
 */
void BrowserHandler::init() {

    // Get socket
    serverSock = getSocket(false, browserPort);

    // Indicate we are willing to accept clients
    if (listen(serverSock, 50) < 0) {
        perror("Listen Error");
        exit(EXIT_FAILURE);
    }

    // Set multiple connections option to true
    int opt = true;
    if (setsockopt(serverSock, SOL_SOCKET, SO_REUSEADDR,
            (char *) &opt, sizeof (opt)) != 0) {
        perror("SetSockOpt Error");
        exit(EXIT_FAILURE);
    }
}

/**
 * Handle one browser query and response
 */
void BrowserHandler::handleComm() {

    // Socket for this client
    int clientSock;

    // Wait for client to accept
    if ((clientSock = accept(serverSock, (struct sockaddr *) NULL,
            NULL)) < 0) {
        perror("Accept Error");
        exit(EXIT_FAILURE);
    }

    // Get response from browser
    String resp = readLine(clientSock);

    // Process response
    String html = processResp(resp);

    // Send reply to browser
    send(clientSock, html.c_str(), html.length(), 0);

    // Give browser time
    sleep(1);

    // Shutdown to release ports
    shutdown(clientSock, SHUT_RDWR);

    // Close connection
    close(clientSock);
}

/**
 * Process a browser response and return the output
 */
String BrowserHandler::processResp(String queryline) {

    // Get destination station name
    vector<String> parts = split(queryline, "/");
    String tempS = strReplace(parts[1], " HTTP", "");
    String destName = strReplace(tempS, "?to=", "");

    // Holder
    String html = "";

    // Add HTTP header
    html += "\nHTTP/1.1 200 OK";
    html += "\nContent-Type: text/html";
    html += "\nConnection: Closed";
    html += "\n";

    // Add HTML payload
    html += "\n<html>";
    html += "\n<body>";
    html += "\n<br>";
    html += "\n<h1> " + stName + " </h1>";
    html += "\n<br>";
    html += "\n<h2> Answer: </h2>";

    // If query is invalid
    if (strEquals(destName, "") || strContains(destName, "favicon")) {

        // Notify
        html += "\n<p style=\"font-size:18px\"><i>";
        html += "No query made";
        html += "</i></p>";
    } else {

        // Else if query is valid
        // Get a result
        String result = timetable.getRoute(destName);

        // Check
        if (strContains(result, "UNAVAIL")) {
            result = "Sorry, no route available today.";
            result += " Refreshing the page may work.";
        }

        // Package up result
        html += "\n<p><b> Trip from " + stName;
        html += " to " + destName;
        html += ": </b></p>";
        html += "\n<p style=\"font-size:18px\"><i>";
        html += result;
        html += "</i></p>";
    }

    html += "\n<br>";
    html += "\n<h3> Other Information: </h3>";
    html += "\n<p><b> Web browser TCP/IP port:</b> ";
    html += to_string(browserPort);
    html += "</p>";
    html += "\n<p><u> Station UDP/IP port:</u>     ";
    html += to_string(stPort);
    html += "</p>";
    html += "\n<p><u> Adjacent station UDP/IP ports:</u> ";
    for (int port : adjPorts) {
        html += to_string(port) + "  ";
    }
    html += "</p>";
    html += "\n<p><b> Random Number:</b> ";
    srand(time(NULL) * time(NULL) * rand());
    html += to_string((rand() % 100) + 1);
    html += "</p>";
    html += "\n</body>";
    html += "\n</html>";

    // Return String
    return html;
}

/**
 * Read line from socket
 */
String BrowserHandler::readLine(int socketfd) {
    int buf_size = 0;
    int in_buf = 0;
    int ret;
    char ch;
    char* buffer = NULL;
    char* new_buffer;

    // Loop
    while (true) {
        // Get char
        ret = read(socketfd, &ch, 1);

        // If end of line reached
        if (ch == '\n') {
            // Stop
            break;
        }

        // Check if more memory needed
        if ((buf_size == 0) || (in_buf == buf_size)) {
            buf_size += 128;
            new_buffer = (char*) realloc(buffer, buf_size);
            buffer = new_buffer;
        }

        buffer[in_buf] = ch;
        ++in_buf;
    }

    // Handle line endings
    if ((in_buf > 0) && (buffer[in_buf - 1] == '\r'))
        --in_buf;

    // Allocate more memory if needed
    if ((buf_size == 0) || (in_buf == buf_size)) {
        ++buf_size;
        new_buffer = (char*) realloc(buffer, buf_size);
        buffer = new_buffer;
    }

    // Terminate
    buffer[in_buf] = '\0';

    // Return String
    return String(buffer);
}

// ----
// ------------------------------------
// ----

/**
 * Constructor
 */
BrowserHandler::BrowserHandler() {
}

/**
 * Special
 * @param orig
 */
BrowserHandler::BrowserHandler(const BrowserHandler & orig) {
}

/**
 * Destructor
 */
BrowserHandler::~BrowserHandler() {
}

File: CITS3002 - Networks/Project - Transport Station - Client and Server/JavaProject/src/comm/BrowserHandler.java
package comm;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Timer;
import java.util.TimerTask;
import java.util.logging.Level;
import java.util.logging.Logger;
import main.Globals;

/**
 * Handles browser interface communications (via TCP)
 *
 * @author David
 */
public class BrowserHandler {

    // Web browser socket server
    private final ServerSocket browserSS;

    /**
     * Create browser handler
     *
     * @throws java.io.IOException
     */
    public BrowserHandler() throws IOException {

        // Initialize socket server to listen for browser connection on port
        browserSS = new ServerSocket(Globals.browserPort);
    }

    /**
     * Process a browser query
     *
     * @param queryline
     * @return
     */
    private String processQuery(String queryline) {

        // Get destination station name
        String[] parts = queryline.split("/");
        String tempS = parts[1].replace(" HTTP", "");
        String destName = tempS.replace("?to=", "");

        // Holder
        String midHtml = "";

        // If query is invalid
        if (destName.equals("") || destName.contains("favicon")) {

            // Notify 
            midHtml += "\n<p style=\"font-size:18px\"><i>";
            midHtml += "No query made";
            midHtml += "</i></p>";
        } else {

            // Else if query is valid
            // Get route from timetable
            String route = Globals.timetable.getRoute(destName);

            // If route has unavailable section
            if (route.contains("UNAVAIL")) {
                route = "Sorry, no route available today.";
                route += " Refreshing the page may work.";
            }

            // Give route heading and stylize
            midHtml += "\n<p><b> Trip from " + Globals.stName;
            midHtml += " to " + destName;
            midHtml += ": </b></p>";
            midHtml += "\n<p style=\"font-size:18px\"><i>";
            midHtml += route;
            midHtml += "</i></p>";
        }

        // Return string
        return midHtml;
    }

    /**
     * Send HTML content to the browser
     *
     * @param socket
     * @param middleContent
     */
    private void sendToBrowser(Socket socket, String middleContent) {

        // Holder
        String full = "";

        // Add HTTP header
        full += "\nHTTP/1.1 200 OK";
        full += "\nContent-Type: text/html";
        full += "\nConnection: Closed";
        full += "\n";

        // Add HTML payload
        full += "\n<html>";
        full += "\n<body>";
        full += "\n<br>";
        full += "\n<h1> " + Globals.stName + " </h1>";
        full += "\n<br>";
        full += "\n<h2> Answer: </h2>";

        // Add content
        full += middleContent;

        // Add bottom statistics
        full += "\n<br>";
        full += "\n<h3> Other Information: </h3>";
        full += "\n<p><b> Web Browser TCP/IP Port:</b> ";
        full += Globals.browserPort;
        full += "</p>";
        full += "\n<p><u> Station UDP/IP Port:</u>     ";
        full += Globals.stPort;
        full += "</p>";
        full += "\n<p><u> Adjacent Station UDP/IP Ports:</u> ";
        full += Globals.adjacentPorts;
        full += "</p>";
        full += "\n<p><b> Random Number:</b> ";
        full += (int) (Math.random() * 100);
        full += "</p>";
        full += "\n</body>";
        full += "\n</html>";

        // Get bytes
        byte[] bytes = full.getBytes();

        // Write to socket
        try {
            socket.getOutputStream().write(bytes);
        } catch (IOException ex) {
            Logger.getLogger(BrowserHandler.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    /**
     * Receive browser connection requests and handle responses
     */
    public void receiveRequest() {

        // Wait to accept connection (as socket) from browser
        try (Socket socket = browserSS.accept()) {

            // When connection has been made:
            // Get browser query
            String query = getFromBrowser(socket);

            // Process query
            String result = processQuery(query);

            // Send result back to browser
            sendToBrowser(socket, result);

        } catch (IOException ex) {
            Logger.getLogger(BrowserHandler.class.getName()).
                    log(Level.SEVERE, null, ex);
        }

        // Do again after a short delay
        new Timer().schedule(new TimerTask() {
            @Override
            public void run() {
                receiveRequest();
            }
        }, 10);
    }

    /**
     * Get the the first line of the browser's HTTP response
     *
     * @param socket
     * @return
     * @throws IOException
     */
    private String getFromBrowser(Socket socket)
            throws IOException {

        // Get input stream and reader
        InputStream is = socket.getInputStream();
        InputStreamReader isr = new InputStreamReader(is);

        // Create reader
        BufferedReader reader = new BufferedReader(isr);

        // Return first line
        return reader.readLine();
    }
}
